rollup/ccc/logger.go

```
@@ -0,0 +1,865 @@
+		l.ecPairingCount++
+		outputLen = 32
+	case common.BytesToAddress([]byte{9}): // &blake2FDisabled{},
+	}
+	l.logCopy(2 * outputLen)
+}
+
+// logCall logs call to a given address, regardless of the address being a precompile or not
+func (l *Logger) logCall(to common.Address, inputLen uint64, inputFn func(int64, int64) ([]byte, error)) {
+	l.logBytecodeAccessAt(to)
+	l.logPrecompileAccess(to, inputLen, inputFn)
+}
+
+func (l *Logger) logCopy(len uint64) {
+	l.copyUsage += len * 2
+}
+
+func (l *Logger) logSha256(inputLen uint64) {
+	const blockRows = 2114
+	const blockSizeInBytes = 64
+	const minPaddingBytes = 9
+
+	numBlocks := (inputLen + minPaddingBytes + blockSizeInBytes - 1) / blockSizeInBytes
+	l.sha256Usage += numBlocks * blockRows
+}
+
+func (l *Logger) CaptureStart(env *vm.EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
+	l.currentEnv = env
+	l.isCreate = create
+	if !l.isCreate {
+		l.logCall(to, uint64(len(input)), func(argOffset, argLen int64) ([]byte, error) {
+			return input[argOffset : argOffset+argLen], nil
+		})
+	} else {
+		l.logRawBytecode(input) // init bytecode
+	}
+
+	if !env.TxContext.IsL1MessageTx {
+		l.sigCount++
+		l.l2TxnsRlpSize += uint64(env.TxContext.TxSize)
+	}
+	l.keccakUsage += computeKeccakRows(uint64(env.TxContext.TxSize))
+	l.keccakUsage += computeKeccakRows(64) // ecrecover per txn
+}
+
+func (l *Logger) CaptureState(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, err error) {
+	if err != nil {
+		return
+	}
+
+	l.evmUsage += evmUsagePerOpCode[op]
+	l.stateUsage += stateUsagePerOpCode[op](scope, depth)
+
+	getInputFn := func(inputOffset int64) func(int64, int64) ([]byte, error) {
+		return func(argOffset, argLen int64) ([]byte, error) {
+			input, err := tracers.GetMemoryCopyPadded(scope.Memory, inputOffset+argOffset, argLen)
+			if err != nil {
+				log.Warn("failed to read call input", "err", err)
```

Now that you reduced `memoryPadLimit back to its original value, will we run into this error in practice? I guess it's not likely that this happens in `logPrecompileAccess`? But if it does, what impact will it have (since `logPrecompileAccess` just ignores the error)?

---

> But if it does, what impact will it have

We might underestimate `copy` circuit usage and cause a reorg.